[incubator/couchdb] Add a chart for Apache CouchDB

kocolosk opened the pull request:
This PR adds a chart that deploys [Apache CouchDB](https://couchdb.apache.org/) as a StatefulSet. It defaults to using the image from the [official Docker repository](https://hub.docker.com/_/couchdb/) as the primary container in each Pod, and attaches a second [sidecar container](https://hub.docker.com/r/kocolosk/couchdb-statefulset-assembler/) to help the Pods in the StatefulSet discover and connect to one another. Some other notable features include:

- Use of a Secret to supply initial admin username and password
- Ability to customize the server configuration through a ConfigMap
- Optional use of PVCs for persistence

I've tried to adhere to the contribution guidelines but do let me know if you see anything amiss.

kocolosk committed:
Add scaffolding for CouchDB chart

kocolosk committed:
Move to a StatefulSet rather than a Deployment

kocolosk committed:
Configure a headless service

This configuration is necessary to ensure that the individual Pods
have well-defined hostnames that are discoverable via DNS.

kocolosk committed:
Make the minimal set of changes for a StatefulSet

Lots of these changes need to be properly parameterized, but first
let's get the basics working.

kocolosk committed:
Fix the default variables for a 3 node cluster

kocolosk committed:
Support generic Erlang VM flags as Values

kocolosk committed:
Specify the location of sidecar image in values

kocolosk committed:
Allow to override the podManagementPolicy

kocolosk committed:
Ensure exact match between labels and selector

kocolosk committed:
Simplify headless service definition

The current version of the sidecar container expects the headless
service to export SRV records for a _couchdb service, so there's no
point in parameterizing the name here. In theory the port could be
omitted, but there are a few unresolved issues with that configuration
as of v1.7 so we just put in the stock CouchDB port number.

kocolosk committed:
Do not pretend that we configure the CouchDB port

In a future version we may be able to pass the port configuration to
CouchDB using a ConfigMap, but for now the port on which the Pods are
listening for HTTP requests is fixed.

kocolosk committed:
Enable configuration of PVCs for the statefulset

kocolosk committed:
Choose a more user-friendly name for spec.replicas

Users familiar with CouchDB may wonder if "replicaCount" refers to the
desired number of replicas per document, as opposed to the size of the
cluster.

kocolosk committed:
Clean up comments, improve example resource limit

kocolosk committed:
Improve the Service / Ingress configuration

This re-adds a traditional Service alongside the headless one so
clients have a stable way to access the CouchDB cluster. It also
refines some of the documentation around the Ingress resource.

kocolosk committed:
Update the chart metadata

kocolosk committed:
Add a basic README

kocolosk committed:
Make some small doc changes to align w/ standards

kocolosk committed:
Enable CouchDB configuration through ConfigMap

kocolosk committed:
Configure admin account, cookie auth using Secrets

kocolosk committed:
Tell the user to run finish_cluster after startup

kocolosk committed:
Tweak the notes to use interactive password prompt

kocolosk commented:
Hopefully the CLA is sorted:

<img width="945" alt="screen shot 2017-12-07 at 4 29 42 pm" src="https://user-images.githubusercontent.com/16679/33739685-fa5aa05c-db6b-11e7-9674-dec8d89040b4.png">

kocolosk commented:
Hi folks, hope you all had a blast down in Austin. Let me know if there's anything else I can do here to jumpstart the review process.

kocolosk commented:
I see you all are swamped with PRs (good problem to have!), just figured I'd bump this one again to see if there's anything you'd want done.

cmluciano commented:
/ok-to-test

cmluciano commented:
/assign @mattfarina @foxish

cmluciano commented on incubator/couchdb/README.md:
````diff
@@ -0,0 +1,93 @@
+First, create a Secret containing `adminUsername`, `adminPassword` and `cookieAuthSecret` keys:
+
+```bash
+$ kubectl create secret generic my-release-couchdb --from-literal=adminUsername=foo --from-literal=adminPassword=bar --from-literal=cookieAuthSecret=baz
````
you might need to add a test to generate these for CI to function properly

kocolosk replied:
Thanks for taking a look! Not sure exactly what you mean by adding a test here. I see one pattern of programmatically generating default Secrets and allowing an override from `values.yaml`, e.g.

https://github.com/kubernetes/charts/blob/master/stable/mongodb/templates/secrets.yaml

Is that what you'd recommend?

kocolosk replied:
OK this is done in 766c84c

bgehman commented on incubator/couchdb/templates/statefulset.yaml:
```diff
@@ -0,0 +1,123 @@
+{{ toYaml .Values.resources | indent 12 }}
+          volumeMounts:
+          - name: database-storage
+            mountPath: /var/lib/couchdb
```
Shouldn't the mountPath be: `/opt/couchdb/data` ?

Per the docs at: https://hub.docker.com/_/couchdb/
`CouchDB uses /opt/couchdb/data to store its data, and is exposed as a volume.`

kocolosk replied:
Good catch! That's a change from 1.x. Will fix.

kocolosk committed:
Fix the database-storage mount point

kocolosk committed:
Automatically create admin account secrets

This change enables the chart to be automically installed in a secure
fashion with a single command.

mattfarina commented on incubator/couchdb/Chart.yaml:
```diff
@@ -0,0 +1,13 @@
+  - https://github.com/apache/couchdb
+  - https://github.com/kocolosk/couchdb-statefulset-assembler
+maintainers:
+  - name: Adam Kocoloski
```
Can you please use your github username instead of your real name. This makes it easier for us to contact you when it comes to continued maintenance of the chart.

mattfarina commented on incubator/couchdb/templates/statefulset.yaml:
```diff
@@ -0,0 +1,123 @@
+apiVersion: apps/v1beta1
```
Can you update the api version to `apps/v1beta2`. Due to the deprecation policy, `apps/v1beta1` will be removed from support in the near future.

kocolosk replied:
OK. Is there any approach to supporting kube versions < 1.8 in that case?

mattfarina commented on incubator/couchdb/values.yaml:
```diff
@@ -0,0 +1,107 @@
+## If allowAdminParty is enabled the cluster will start up without any database
+## administrator account; i.e., all users will be granted administrative
+## access. Otherwise, the system will look for a Secret called
+## <ReleaseName>-couchdb containing `adminUsername`, `adminPassword` and
```
The name of the secret being used is a hard coded pattern. What about letting users specify a name. If that name isn't supplied a default pattern is used. Being able to set the name of the secret is a pattern we've been using in other charts.

kocolosk replied:
Happy to make this change, I found the current pattern somewhat awkward as I was working through it. Are you suggesting to make the name of the Secret created by the chart configurable, or to enable the `secretKeyRef` in the StatefulSet to be configurable?

I grepped around the chart repo looking for an example where someone had done this, but the first dozen or so I looked at all used the current pattern in this PR. Do you have one in mind that implements what you're suggesting so I can use it as a model?

mattfarina reviewed (commented):
Thanks you for submitting the PR for this chart. I had a little initial feedback.

kocolosk committed:
Use GitHub username for maintainer name

kocolosk committed:
Target latest StatefulSet API

This also means we require Kube 1.8.

svnlto commented:
Patiently waiting for an update on this.

mattfarina commented:
/lgtm

@kocolosk In the metadata for your assembler image it may be worth linking back to the source on GitHub to make it easier for folks to find it.
